\documentclass[a4paper]{article}\usepackage{titling}\newcommand{\subtitle}[1]{% \posttitle{%  \par\end{center}  \begin{center}\large#1\end{center}  \vskip0.5em}%}\usepackage{helvet}\usepackage[helvet]{sfmath}\renewcommand{\familydefault}{\sfdefault}\usepackage[american]{babel}\usepackage{hyperref}\usepackage{a4wide}\usepackage{color}%\usepackage{longtable}%\usepackage{multirow}%\usepackage{rotating}\usepackage{graphicx}\usepackage{authblk}\usepackage[square,numbers]{natbib}\usepackage{amsmath}%\usepackage[top=1in, bottom=1in, left=1in, right=1in]{geometry}%maths\DeclareMathOperator*{\argmax}{argmax}\DeclareMathOperator*{\argmin}{argmin}%color definitions\definecolor{MyDarkRed}{rgb}{0.5,0.05,0}\definecolor{MyDarkBlue}{rgb}{0,0.05,0.5}\definecolor{darkred}{rgb}{0.55,0,0}\definecolor{darkblue}{rgb}{0,0,0.55}\definecolor{darkgreen}{rgb}{0,0.39,0}\definecolor{darkviolet}{rgb}{0.58,0,0.83}\definecolor{darkorange}{rgb}{1,0.55,0}\definecolor{darkyellow}{rgb}{1,0.89,0}%graphics path\graphicspath{{saveres/}}\begin{document}\title{{\Huge{Supplementary Information}}\vspace{3cm}}\subtitle{{\LARGE{\textbf{Exploring consistency between two large pharmacogenomic datasets}}}\vspace{1cm}}\author{Benjamin Haibe-Kains, Nehme El-Hachem, Nicolai Juul Birkbak, Andrew C. Jin, Andrew H. Beck, Hugo J.W.L. Aerts, John Quackenbush}%Benjamin Haibe-Kains1,2,$, Nehme Hachem1, ……. Nicolai Juul Birkbak, Andrew H. Beck,….. Hugo J.W.L. Aerts3,4, John Quackenbush3,5%1 Bioinformatics and Computational Genomics Laboratory, Institut de recherches cliniques de Montréal, University of Montreal, Montreal, Quebec, Canada, 2 Division of Experimental Medicine, McGill University, Montreal, Quebec, Canada, 3 Department of Biostatistics and Computational Biology and Center for Cancer Computational Biology, Dana-Farber Cancer Institute, Boston, MA, USA, 4 Department of Radiation Oncology, Dana-Farber Cancer Institute, Brigham and Women’s Hospital Boston, MA, USA, 5 Department of Cancer Biology, Dana-Farber Cancer Institute, Boston, MA, USA%$ Corresponding author\date{}\maketitle\newpage\tableofcontents%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\clearpage \section{Full Reproducibility of the Analysis Results} We will describe how to fully reproduce the figures and tables reported in the main manuscript. We automated the analysis pipeline so that minimal manual interaction is required to reproduce our results. To do this, one must simply:\begin{enumerate}\item Set up the software environment\item Run the R scripts\item Generate the Supplementary Information\end{enumerate}\subsection{Set up the software environment}We developed and tested our analysis pipeline using R running on linux and Mac OSX platforms.\bigskip\noindent To mimic our software environment the following R packages should be installed:\input{sessionInfoR.tex}All these packages are available on CRAN\footnote{\href{http://cran.r-project.org}{http://cran.r-project.org}} or Bioconductor\footnote{\href{http://www.bioconductor.org}{http://www.bioconductor.org}}, except for jetset which is available on the CBS website\footnote{\href{http://www.cbs.dtu.dk/biotools/jetset/}{http://www.cbs.dtu.dk/biotools/jetset/}}.%except for the package hthgu133afrmavecs which is provided by the InSilicoDB team\footnote{\href{https://insilicodb.org}{https://insilicodb.org}} (also provided in Supplementary File for convenience, see below).\bigskip\noindent Run the following commands in a R session to install all the required packages:\begin{verbatim}source("http://bioconductor.org/biocLite.R")biocLite(c("AnnotationDbi", "affy", "affyio", "hthgu133acdf",  "hthgu133afrmavecs", "hgu133plus2cdf", "hgu133plus2frmavecs",  "org.Hs.eg.db", "genefu", "biomaRt", "frma", "Hmisc", "vcd",  "epibasix", "amap", "gdata", "WriteXLS", "xtable", "plotrix",  "R.utils", "DBI", "GSA", "gplots"))\end{verbatim}\noindent Note that you may need to install Perl\footnote{\href{http://www.perl.org/get.html}{http://www.perl.org/get.html}} and its module Text::CSV\_XS for the WriteXLS package to write xls file; once Perl is installed in your system, use the following command to install the Text::CSV\_XS module through CPAN\footnote{\href{http://www.cpan.org/modules/INSTALL.html}{http://www.cpan.org/modules/INSTALL.html}}:\begin{verbatim}cpan Text/CSV_XS.pm\end{verbatim}\noindent Lastly, follow the instructions on the CBS website to properly install the jetset package or use the following commands in R:\begin{verbatim}download.file(url="http://www.cbs.dtu.dk/biotools/jetset/current/jetset_1.4.0.tar.gz",  destfile="jetset_1.4.0.tar.gz")install.packages("jetset_1.4.0.tar.gz", repos=NULL, type="source")\end{verbatim}\bigskip\noindent Once the packages are installed, uncompress the archive (\texttt{CDRUG.zip}) provided as \textbf{Supplementary File~2} accompanying the manuscript. This should create a directory on the file system containing the following files:\begin{description}\item[\texttt{CDRUG\_foo.R}] Script containing the definitions of all functions required for the analysis pipeline.\item[\texttt{CDRUG\_normalization\_cgp.R}] Script to curate, annotate and normalize of CGP data.\item[\texttt{CDRUG\_normalization\_ccle.R}] Script to curate, annotate and normalize of CCLE data.\item[\texttt{CDRUG\_normalization\_gsk.R}] Script to curate, annotate and normalize of GSK data.\item[\texttt{CDRUG\_format.R}] Script to identify common cell lines, tissue types and drugs investigated both in CGP and CCLE.\item[\texttt{CDRUG\_analysis.R}] Script generating all the figures and tables reported in the manuscript.\item[\texttt{CDRUG\_analysisbis.R}] Script generating Figure~4 in the manuscript.\item[\texttt{CDRUG\_analysisbis\_gsk.R}] Script comparing the IC$_{50}$ measures between CGP, CCLE and GSK.\item[\texttt{CDRUG\_pipeine.R}] Master script running all the scripts listed above to generate the analysis results.\item[\texttt{gsea2-2.0.13.jar}] GSEA java executable; it can also be downloaded from the GSEA website\footnote{\href{http://www.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/software/gsea2-2.0.13.jar}{http://www.broadinstitute.org/gsea/msigdb/download\_file.jsp?filePath=/resources/software/gsea2-2.0.13.jar}}.\item[\texttt{c5.all.v4.0.entrez.gmt}] Definition of genesets based on Entrez Gene IDs; it can also be downloaded from the GSEA website\footnote{\href{http://www.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/4.0/c5.all.v4.0.entrez.gmt}{http://www.broadinstitute.org/gsea/msigdb/download\_file.jsp?filePath=/resources/msigdb/4.0/c5.all.v4.0.entrez.gmt}}.\item[\texttt{matching\_cell\_line\_CCLE\_CGP.csv}] Curation of cell line name to match CGP and CCLE nomenclatures.\item[\texttt{matching\_tissue\_type\_CCLE\_CGP.csv}] Curation of tissue type name to match CGP and CCLE nomenclatures.\item[\texttt{matching\_cell\_line\_GSK\_CCLE\_CGP.csv}] Curation of cell line name to match those of GSK with those of CGP and CCLE.\item[\texttt{matching\_tissue\_type\_GSK\_CCLE\_CGP.csv}] Curation of tissue type name to match those of GSK with those of CGP and CCLE.\item[\texttt{cdrug\_suppl\_info.tex}] The \LaTeX \ file of the present supplementary information\end{description}\bigskipAll the files required to run the automated analysis pipeline are now in place. It is worth noting that raw gene expression and drug sensitivity data are voluminous, please ensure that at least 25GB of storage are available.\subsection{Run the R scripts}Open a terminal window and go to the \texttt{CDRUG} directory. You can easily run the analysis pipeline either in batch mode or in a R session. Before running the pipeline you can specify the number of CPU cores you want to allocate to the analysis (by default only 1 CPU core will be used). To do so, open the script \texttt{CDRUG\_pipeline.R} and update line \#33:\begin{verbatim}nbcore <- 4\end{verbatim}to allocate four CPU cores for instance.\bigskip\noindent To run the full pipeline in batch mode, simply type the following command:\texttt{R CMD BATCH CDRUG\_pipeline.R Rout \&}\noindent The progress of the pipeline could be monitored using the following command:\texttt{tail -f Rout}\bigskip\noindent To run the full analysis pipeline in an R session, simply type the following command:\texttt{source("CDRUG\_pipeline.R")}\noindent Key messages will be displayed to monitor the progress of the analysis.\bigskip\noindent The analysis pipeline was developed so that all intermediate analysis results are saved in the directories \texttt{data} and \texttt{saveres}. Therefore, in case of interruption, the pipeline will restart where it stopped.\subsection{Generate the Supplementary Information}After completion of the analysis pipeline a directory \texttt{saveres} will be created to contain all the intermediate results, tables and figures reported in the main manuscript and this Supplementary Information.%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\newpage\section{Supplementary Methods}\subsection{Data retrieval and curation}\subsubsection{CGP (release 4, March 2013)}Gene expression, mutation data and cell line annotations were downloaded from \href{ftp://ftp.ebi.ac.uk//pub/databases/microarray/data/experiment/MTAB/E-MTAB-783/}{ArrayExpress}. Drug sensitivity measurements and drug information were downloaded from the CGP website (\textcolor{MyDarkBlue}{\href{ftp.sanger.ac.uk/pub4/cancerrxgene/releases/release-2.0/}{CGP}}) and the Nature website (\textcolor{MyDarkBlue}{\href{http://www.nature.com/nature/journal/v483/n7391/full/nature11005.html}{Nature article}}), respectively.Minimum and maximum screening concentrations for each drug/cell line were extracted from \texttt{gdsc\_compounds\_conc\_w2.csv} ($\mu$M).The natural logarithm of IC$_{50}$ measurements were retrieved from \texttt{gdsc\_manova\_input\_w2.csv} in column \texttt{"*\_IC\_50"} (referred to as $x$) and subsequently transformed using $-\log_{10}(\exp(x))$; high values are representative of cell line sensitivity to drugs.The AUC measurements were retrieved from \texttt{gdsc\_manova\_input\_w2.csv} in column \texttt{"*\_AUC"} (referred to as $x$); high values are representative of cell line sensitivity to drugs.\subsubsection{CCLE (release March 2013)}Gene expression, mutation data cell line annotations and drug information were downloaded from the CCLE website (\textcolor{MyDarkBlue}{\href{http://www.broadinstitute.org/ccle/data/browseData?conversationPropagation=begin}{CCLE}}) Drug sensitivity data were downloaded from the Nature website (\textcolor{MyDarkBlue}{\href{http://www.nature.com/nature/journal/v492/n7428/full/nature11735.html}{Nature letter addendum}}); Screening concentrations for each drug/cell line were extracted from Supplementary Table 11 in column E ($\mu$M).IC$_{50}$ measurements were retrieved from Supplementary Table 11 in column J ("\texttt{IC$_{50}$ $\mu$M (norm)}") (referred to as $x$) and subsequently transformed into logarithmic scale, $-\log_{10}(x)$; high values are representative of cell line sensitivity to drugs.AUC measurements were retrieved from Supplementary Table 11 in column L ("\texttt{ActArea (norm)}") and subsequently divided by the number of drug concentrations tested (8); high values are representative of cell line sensitivity to drugs.\subsection{Cell line annotations}Cell line names were harmonized in both CGP and CCLE to match identical cell lines; this was done through manual search over alternative names of cell lines, as reported in CGP and CCLE cell line annotation files and online databases such as \textcolor{MyDarkBlue}{\href{http://bioinformatics.istge.it/hypercldb/}{hyperCLDB}} and \textcolor{MyDarkBlue}{\href{http://bioinfoweb.com/}{BioInformationWeb}}. We identified 471 cancer cell lines being investigated both in CGP and CCLE:\begin{center}\includegraphics[keepaspectratio=true,width=0.4\textwidth]{intersection_cellines_cgp_ccle_paper}\end{center}Tissue type nomenclature from CGP \citep{Garnett:2012fc} was chosen throughout this study, CCLE tissue type information \citep{Barretina:2012fp} was therefore updated to follow this nomenclature, which resulted in 24 tissue types:\medskip{\small{\begin{center}\input{saveres/tissue_type_cgp_ccle_paper.tex}\end{center}}}\medskipAll the curation steps have been documented in the scripts \texttt{CDRUG\_normalization\_cgp.R}, \texttt{CDRUG\_normalization\_ccle.R}, and \texttt{CDRUG\_format.R}.\subsection{Mutation data}We focused our analyses on missense mutations identified both in CGP and CCLE. For CGP, coding variants in 68 genes were extracted from \textcolor{MyDarkBlue}{\href{ftp://ftp.sanger.ac.uk/pub4/cancerrxgene/releases/release-2.0/gdsc_manova_input_w2.csv}{gdsc\_manova\_input\_w2.csv}}. For CCLE, coding variants in 1667 genes (column '\texttt{Protein\_Change}') measured using the Oncomap3 and hybrid capture platforms were extracted from \textcolor{MyDarkBlue}{\href{http://www.broadinstitute.org/ccle/downloadFile/DefaultSystemRoot/exp_10/ds_23/CCLE_Oncomap3_2012-04-09.maf?downloadff=true&fileId=3000}{CCLE\_Oncomap3\_2012-04-09.maf}} and \\\textcolor{MyDarkBlue}{\href{http://www.broadinstitute.org/ccle/downloadFile/DefaultSystemRoot/exp_10/ds_26/CCLE_hybrid_capture1650_hg19_NoCommonSNPs_NoNeutralVariants_CDS_2012.05.07.maf.gz?downloadff=true&fileId=6873}{CCLE\_hybrid\_capture1650\_hg19\_NoCommonSNPs\_NoNeutralVariants\_CDS\_2012.05.07.maf.}}, respectively.\subsection{Correlation between mutation profiles}For the 471 cell lines investigated in both CGP and CCLE, we identified 64 genes whose missense mutations in protein coding genes have been measured in both studies. We followed the approach used in the original studies by transforming the mutation data into binary matrices reporting the presence or absence of at least one mutation in a given gene and cell line. The agreement between mutation profiles measured in CGP and CCLE was then measured using Cohen's $\kappa$ statistics. We used the following qualitative descriptions of $\kappa$ values associated with intervals: $\kappa < 0.5$, poor agreement; $0.5 \leq \kappa < 0.6$, fair agreement; $0.6 \leq \kappa < 0.7$, moderate agreement; $0.7 \leq \kappa < 0.8$, substantial agreement; and $\kappa \geq 0.8$, almost perfect agreement. \subsection{Correlation between drug sensitivity measures}We assessed consistency between drug sensitivity measures (IC$_{50}$ and AUC), gene-drug and pathway-drug association coefficients in CGP and CCLE by computing Spearman rank-ordered correlations when $\geq 10$ measures were available. Typically qualitative descriptions of correlation coefficients are associated with intervals: $r_s < 0.5$, poor correlation; $0.5 \leq r_s < 0.6$, fair correlation; $0.6 \leq r_s < 0.7$, moderate correlations; $0.7 \leq r_s < 0.8$, substantial correlation; and $r_s \geq 0.8$, almost perfect correlation. \subsection{Drug sensitivity calling}To categorize cell line sensitivity into three categories (resistant, intermediate and sensitivity) we used the \textit{waterfall} approach described in the CCLE study \citep{Barretina:2012fp}. The full procedure, as communicated by Dr. Kavitha Venkatesan (personal communication) is described below:\begin{enumerate}\item Extract the drug sensitivity measurements, either IC$_{50}$ or AUC\item Sort log IC$_{50}$ values (or AUC) of the cell lines to generate a \textit{waterfall} distribution.\item If the waterfall distribution is non-linear (Pearson correlation coefficient to the linear fit $\leq 0.95$), estimate the major inflection point of the log IC$_{50}$ curve as the point on the curve with the maximal distance to a line drawn between the start and end points of the distribution.\item If the waterfall distribution appears linear (Pearson correlation coefficient to the linear fit $> 0.95$), then use the median IC$_{50}$ instead.\item Cell lines within a 4-fold IC$_{50}$ (or within a 1.2-fold AUC) difference centered around this inflection point are classified as being \textit{intermediate}, cell lines with lower IC$_{50}$ (or AUC) values than this range are defined as \textit{sensitive}, and those with IC$_{50}$ (or AUC) values higher than this range are called “\textit{resistant}.\item Require at least $x=5$ sensitive and $x=5$ resistant cell lines after applying these criteria.\end{enumerate}\subsection{Consistency between drug sensitivity calls}We assessed the consistency of drug sensitivity calling, as computed using the waterfall method described in the CCLE study \citep{Barretina:2012fp}, using Cohen’s Kappa ($\kappa$) coefficient. Similarly to correlation coefficients, qualitative descriptions of Kappa coefficients are associated with intervals: $\kappa < 0.2$, poor agreement; $0.2 \leq \kappa < 0.4$, fair agreement; $0.4 \leq \kappa < 0.6$, moderate agreement; $0.6 \leq \kappa < 0.8$, substantial agreement; and $\kappa \geq 0.8$, almost perfect agreement.\subsection{Gene-drug associations}We assessed the association between gene expression and drug response, referred to as gene-drug association, using a linear regression model controlled for tissue type:$ Y = \beta_0 + \beta_i G_i + \beta_t T $\noindent where $Y$ denote the drug sensitivity variable, $G_i$ and $T$ denote the expression of gene $i$ and the tissue type respectively and $\beta$s are the regression coefficients. The strength of gene-drug association is quantified by $\beta_i$, above and beyond the relationship between drug sensitivity and tissue type. The variables $Y$ and $G$ are scaled (standard deviation equals to 1) to estimate standardized coefficients from the linear model. Significance of the gene-drug association is estimated by the statistical significance of $\beta_i$ ($t$ statistic).\subsection{Pathway-drug associations}For each drug, genes were ranked according to the statistical significance of their gene-drug association ($t$ statistic). We then used this drug-specific gene ranking to perform pre-ranked gene set enrichment analyses (GSEA version 2.0.13 \citep{subramanian2005gene}) to assess enrichment of gene ontology terms curated in MSigDB (\texttt{c5.all.v4.0.entrez.gmt}). Only pathways whose corresponding gene sets contained between 15 genes and 250 genes, were considered for further analysis (913 gene sets). We used the resulting normalized enrichment scores to quantify the strength of pathway-drug associations.\subsection{GSK (release September 2011)}Gene expression data and cell line annotations were downloaded from \textcolor{MyDarkBlue}{\href{https://cabig.nci.nih.gov/community/caArray_GSKdata/?searchterm=None}{caBIG}}. Drug sensitivity measurements and drug information were downloaded from Supplementary Table~2 of \citeauthor{Greshock:2010hf}~\citep{Greshock:2010hf} (\textcolor{MyDarkBlue}{\href{http://cancerres.aacrjournals.org/content/suppl/2010/04/19/0008-5472.CAN-09-3788.DC1/stab_2.xls}{stab\_2.xls}}).\medskip\noindent Cell line names were manually curated and matched to those of CGP and CCLE datasets. AFter curation we found 231 cell lines shared between CGP and GSK, and 249 cell lines shared between CCLE and GSK.\begin{center}\includegraphics[keepaspectratio=true,width=0.4\textwidth]{intersection_cellines_cgp_ccle_gsk_paper}\end{center}\medskip\noindent Curation and data pre-processing are fully documented in \texttt{CDRUG\_normalization\_gsk.R}.%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\newpage\section{Supplementary Tables}\paragraph{Supplementary Table~1} Description of the 15 anticancer drugs screened both in CGP and CCLE studies.\begin{center}\begin{tabular}{| r | p{6cm} | c | l |}\hlineCompound & Target(s)& Class & Organization \\\hline\hlineErlotinib & EGFR	 & Kinase inhibitor & 	Genentech\\\hlineLapatinib & EGFR, HER2	 & Kinase inhibitor & 	GlaxoSmithKline\\\hlinePHA-665752 & c-MET	 & Kinase inhibitor	 & Pfizer \\\hlineCrizotinib & c-MET, ALK	 & Kinase inhibitor	 & Pfizer \\\hlineTAE684	 & ALK	 & Kinase inhibitor & Novartis\\\hlineNilotinib	 & Abl/Bcr-Abl	 & Kinase inhibitor & 	Novartis\\\hlineAZD0530 & Src, Abl/Bcr-Abl, EGFR	 & Kinase inhibitor & AstraZeneca\\\hlineSorafenib	 & Flt3, C-KIT, PDGFRbeta, RET, Raf kinase B, Raf kinase C, VEGFR-1, KDR, FLT4	 & Kinase inhibitor	 & Bayer\\\hlinePD-0332991 & 	CDK4/6	 & Kinase inhibitor & Pfizer \\\hlinePLX4720 & RAF 	 & Kinase inhibitor & 	Plexxikon\\\hlinePD-0325901	 & MEK	 & Kinase inhibitor	 & Pfizer\\ \hlineAZD6244 & MEK & Kinase inhibitor & AstraZeneca\\\hlineNutlin-3 & MDM2 & Other & Roche\\\hline17-AAG & HSP90 & Other & Bristol-Myers Squibb\\\hlinePaclitaxel & beta-tubulin & Cytotoxic & Bristol-Myers Squibb\\\hline\end{tabular}\end{center}\clearpage\paragraph{Supplementary Table~2} Spearman correlation coefficients and significance for (A) IC$_{50}$ measures, (B) all gene-drug associations computed with IC$_{50}$, (C) significant (FDR $<$ 20\%) gene-drug associations computed with IC$_{50}$, (D) all pathway-drug associations computed with IC$_{50}$, (E) significant (FDR $<$ 20\%) pathway-drug associations computed with IC$_{50}$. Significance of positive correlation coefficient is reported using the following convention: '***' for p-value $<0.001$, '**' for p-value $<0.01$, '*' for p-value $<0.05$, 'NS' for p-value $\geq 0.05$. When less than 10 IC$_{50}$ values were available, correlation coefficient was not computed and was therefore represented by empty cells in the table.\input{saveres/correlations_ic50_pvalue_paper.tex}\clearpage\paragraph{Supplementary Table~3} Spearman correlation coefficients and significance for (A) AUC measures, (B) all gene-drug associations computed with AUC, (C) significant (FDR $<$ 20\%) gene-drug associations computed with AUC, (D) all pathway-drug associations computed with AUC, (E) significant (FDR $<$ 20\%) pathway-drug associations computed with AUC. Significance of positive correlation coefficient is reported using the following convention: '***' for p-value $<0.001$, '**' for p-value $<0.01$, '*' for p-value $<0.05$, 'NS' for p-value $\geq 0.05$. When less than 10 AUC values were available, correlation coefficient was not computed and was therefore represented by empty cells in the table.\input{saveres/correlations_auc_pvalue_paper.tex}\clearpage\paragraph{Supplementary Table~4} Contingency tables comparing the sensitivity calls (res, inter, and sens standing for resistant, intermediate and sensitive drug phenotype, respectively) computed from IC$_{50}$ measures for each of the 15 drugs screened both in CGP and CCLE. The Kappa coefficient, its confidence interval and its significance are reported below each contingency table.\begin{center}\includegraphics[keepaspectratio=true,width=1\textwidth]{ic50_sensitivity_calling_intermediate_ccle_cgp_paper}\end{center}\clearpage\paragraph{Supplementary Table~5} Contingency tables comparing the sensitivity calls (res, inter, and sens standing for resistant, intermediate and sensitive drug phenotype, respectively) computed from AUC measures for each of the 15 drugs screened both in CGP and CCLE. The Kappa coefficient, its confidence interval and its significance are reported below each contingency table.\begin{center}\includegraphics[keepaspectratio=true,width=1\textwidth]{auc_sensitivity_calling_intermediate_ccle_cgp_paper}\end{center}%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\newpage\section{Supplementary Figures}%\clearpage\paragraph{Supplementary Figure~1} Box plot of the correlations of gene expression profiles between identical cell lines in CGP and CCLE, across tissue types. Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{ge_var1000_cellines_boxplot2_ccle_cgp_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~2} Box plot of the correlations of missense mutation profiles between identical cell lines in CGP and CCLE. Wilcoxon rank sum test was used to test whether agreement was significantly higher in identical cell lines compared to different cell lines (upper right corner).\begin{center}\includegraphics[keepaspectratio=true,width=0.7\textwidth]{mut_cellines_boxplot_ccle_cgp_paper}\end{center}\clearpage\paragraph{Supplementary Figure~3} Box plot of the correlations of (missense) mutation profiles between identical cell lines in CGP and CCLE, across tissue types. Kruskal-Wallis test was used to test whether agreement significantly depended on tissue type (upper right corner).%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{mut_cellines_boxplot2_ccle_cgp_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~4} Scatter plot reporting the IC$_{50}$ values of Camptothecin for 252 cell lines screened within the CGP project, as measured at the facilities of the Massachusetts General Hospital (MGH) and the Wellcome Trust Sanger Institute (WTSI). Spearman correlation coefficient (R$_s$) is reported in the upper left corner.\begin{center}\includegraphics[keepaspectratio=true,width=0.6\textwidth]{cgp_camptothecin_mgh_wtsi_paper}\end{center}\clearpage\paragraph{Supplementary Figure~5} Scatter plots reporting the drug sensitivity measurements, which are the IC$_{50}$ values within the range of tested concentration (thus excluding extrapolated IC$_{50}$ in CGP and placeholder values in CCLE) in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug where significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. %\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_ic50_cgp_ccle_filtconc_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~6} Scatter plots reporting the drug sensitivity (AUC) measured in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug where significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. %\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_auc_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~7} Box plot of the correlations of the sensitivity measures for 15 drugs, across tissue types. (A) Correlations between IC$_{50}$ measures; (B) correlations between AUC measures. Correlations were estimated using the Spearman coefficient ($R_s$). Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\textbf{A}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_ic50_ccle_cgp_tissue_paper}\\\vspace{-0.75cm}\textbf{B}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_auc_ccle_cgp_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~8} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for drug sensitivity computed with IC$_{50}$ and AUC measures both in CGP and CCLE, across tissue types. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. %\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_ic50_auc_tissue1_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_ic50_auc_tissue2_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~9} Bar plot reporting Cohen's Kappa coefficients (K) quantitatively assessing the concordance between drug sensitivity calls computed with IC$_{50}$ and AUC measures both in CGP and CCLE.\begin{center}\includegraphics[keepaspectratio=true,width=0.6\textwidth]{barplot_ic50_auc_call_cgp_ccle_paper}\end{center}\clearpage\paragraph{Supplementary Figure~10} Scatter plots reporting the gene-drug associations computed with IC$_{50}$, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_ic50_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~11} Scatter plots reporting the gene-drug associations computed with AUC, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_auc_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~12} Box plot of the correlations of the gene-drug associations for the 15 drugs, across tissue types. (A) Correlations between gene-drug associations computed with IC$_{50}$ in CGP and CCLE; (B) correlations between gene-drug associations computed with AUC in CGP and CCLE. Correlations were estimated using the Spearman coefficient ($R_s$). Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\textbf{A}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_assoc_ic50_ccle_cgp_tissue_paper}\\\vspace{-0.75cm}\textbf{B}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_assoc_auc_ccle_cgp_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~13} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for gene-drug associations computed with IC$_{50}$ and AUC measures both in CGP and CCLE, across tissue types. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. If none positive correlations can be computed for a given tissue type, the plot is omitted.\vspace{-0.5cm}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_tissue1_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_tissue2_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_tissue3_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_tissue4_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~14} Scatter plots reporting the significant (FDR$<$20\%) gene-drug associations computed with IC$_{50}$, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_ic50_filt_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~15} Scatter plots reporting the significant (FDR$<$20\%) gene-drug associations computed with AUC, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE; (B) The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_auc_filt_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~16} Box plot of the correlations of the significant (FDR $ < 20\%$) gene-drug associations for the 15 drugs, across tissue types. (A) Correlations between gene-drug associations computed with IC$_{50}$ in CGP and CCLE; (B) correlations between gene-drug associations computed with AUC in CGP and CCLE. Correlations were estimated using the Spearman coefficient ($R_s$). Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\textbf{A}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_assoc_ic50_ccle_cgp_signif_tissue_paper}\\\vspace{-0.75cm}\textbf{B}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_assoc_auc_ccle_cgp_signif_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~17} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for significant (FDR $ < 20\%$) gene-drug associations computed with IC$_{50}$ and AUC measures both in CGP and CCLE, across tissue types. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. If none positive correlations can be computed for a given tissue type, the plot is omitted.\vspace{-0.35cm}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_signif_tissue1_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_signif_tissue2_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_assoc_ic50_auc_signif_tissue3_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~18} Scatter plots reporting the pathway-drug associations computed with IC$_{50}$, as quantified by the enrichment score from gene set enrichment analysis, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_gsea_ic50_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~19} Scatter plots reporting the pathway-drug associations computed with AUC, as quantified by the enrichment score from gene set enrichment analysis, in the 471 cell lines and for each the 15 drugs studied by both CGP and CCLE. The last bar plot (bottom right corner) reports the positive Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_gsea_auc_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~20} Box plot of the correlations of the pathway-drug associations for the 15 drugs, across tissue types. (A) Correlations between pathway-drug associations computed with IC$_{50}$ in CGP and CCLE; (B) correlations between pathway-drug associations computed with AUC in CGP and CCLE. Correlations were estimated using the Spearman coefficient ($R_s$). Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\textbf{A}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_gsea_ic50_ccle_cgp_tissue_paper}\\\vspace{-0.75cm}\textbf{B}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_gsea_auc_ccle_cgp_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~21} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for pathway-drug associations computed with IC$_{50}$ and AUC in CGP and CCLE, across tissue types. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. If none positive correlations can be computed for a given tissue type, the plot is omitted.%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_tissue1_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_tissue2_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_tissue3_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_tissue4_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~22} (A) Scatter plots reporting the significant (FDR$<$20\%) pathway-drug associations computed with IC$_{50}$, as quantified by the enrichment score from gene set enrichment analysis, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the positive Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_gsea_ic50_filt_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~23} (A) Scatter plots reporting the significant (FDR$<$20\%) pathway-drug associations computed with AUC, as quantified by the enrichment score from gene set enrichment analysis, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the positive Spearman correlation coefficient (R$_s$) or each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_gsea_auc_filt_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~24} Box plot of the correlations of the significant (FDR $ < 20\%$) pathway-drug associations for the 15 drugs, across tissue types. (A) Correlations between significant pathway-drug associations computed with IC$_{50}$ in CGP and CCLE; (B) correlations between significant pathway-drug associations computed with AUC in CGP and CCLE. Correlations were estimated using the Spearman coefficient ($R_s$). Kruskal-Wallis test was used to test whether correlations significantly depended on tissue type (upper right corner).%\begin{center}\textbf{A}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_gsea_ic50_ccle_cgp_signif_tissue_paper}\\\vspace{-0.75cm}\textbf{B}\\\includegraphics[keepaspectratio=true,width=1\textwidth]{boxplot2_gsea_auc_ccle_cgp_signif_tissue_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~25} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for significant (FDR $ < 20\%$) pathway-drug associations computed with IC$_{50}$ and AUC in CGP and CCLE, across tissue types. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. If none positive correlations can be computed for a given tissue type, the plot is omitted.\vspace{-0.35cm}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_signif_tissue1_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_signif_tissue2_cgp_ccle_paper}%\end{center}%\begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1\textwidth]{barplot_gsea_ic50_auc_signif_tissue3_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~26} Scatter plots reporting the mutation-drug associations computed with IC$_{50}$, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the positive Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_mut_ic50_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~27} Scatter plots reporting the mutation-drug associations computed with AUC, as quantified by the standardized coefficient of the gene of interest in a linear model controlled for tissue type, in the 471 cell lines and for each the 15 drugs investigated both in CGP and CCLE. The last bar plot (bottom right corner) reports the positive Spearman correlation coefficient (R$_s$) for each drug.%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{scatterbarplot_assoc_mut_auc_cgp_ccle_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~28} Bar plot reporting the positive Spearman correlation coefficients ($R_s$) for the mutation-drug associations computed with IC$_{50}$ and AUC measures both in CGP and CCLE. Significance of each correlation coefficient is reported using the symbol '*' if p-value $< 0.05$. \begin{center}%\hspace{-2cm}\includegraphics[keepaspectratio=true,width=0.7\textwidth]{barplot_assoc_mut_ic50_auc_cgp_ccle_paper}\end{center}\clearpage\paragraph{Supplementary Figure~28} Box plots reporting, for the 15 drugs in the 471 cell lines investigated both in CGP and CCLE, the correlations between the pathway-drug associations with IC$_{50}$ and AUC, as well as the significant (FDR $ < 20\%$) pathway-drug associations with IC$_{50}$ and AUC. Each box represent the datasets used to compute correlations:\begin{itemize}\item 'Original' refers to the original datasets which are [CGP$_g$~+~CGP$_d$]~vs.~[CCLE$_g$~+~CCLE$_d$],\item 'GeneCGP.fixed' refers to [CGP$_g$~+~CGP$_d$]~vs.~[CGP$_g$~+~CCLE$_d$],\item 'GeneCCLE.fixed' refers to [CCLE$_g$~+~CGP$_d$]~vs.~[CCLE$_g$~+~CCLE$_d$],\item 'DrugCGP.fixed' refers to [CGP$_g$~+~CGP$_d$]~vs.~[CCLE$_g$~+~CGP $_d$],\item 'DrugCCLE.fixed' refers to [CGP$_g$~+~CCLE$_d$]~vs.~[CCLE$_g$~+~CCLE$_d$].\end{itemize}where $_g$ and $_d$ stand for gene expressions and drug sensitivities, respectively. Kruskal-Wallis test was used to test whether correlations significantly depended on dataset (upper right corner).%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{boxplot2_gsea_gene_ic50_auc_cgp_ccle_swap_paper}%\end{center}\clearpage\paragraph{Supplementary Figure~30} Scatter plots reporting the drug sensitivity measurements (IC$_{50}$ ) of (A) Lapatinib and (B) Paclitaxel in CGP, CCLE and GSK datasets.\vspace{1cm}\hspace{-1cm}\textbf{A}%\begin{center}\hspace{-2cm}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{ic50_LAPATINIB_ccle_cgp_gsk_paper}\\%\end{center}\hspace{-1cm}\textbf{B}\hspace{-2cm}%\begin{center}\includegraphics[keepaspectratio=true,width=1.2\textwidth]{ic50_PACLITAXEL_ccle_cgp_gsk_paper}%\end{center}%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\newpage\section{List of Abbreviations}\begin{tabular}{rcl}AUC & & Area under the drug sensitivity curve.\\CGP & & Cancer Genome Project initiated by the Wellcome Sanger Institute.\\CCLE & & The Cancer Cell Liners Encyclopedia initiated by Novartis and the Broad Institute.\\$\mathrm{IC_{50}}$ & & Concentration in micro molar [$\mu$M] at which the drug inhibited 50\% of the cellular growth.\\FDR & & False Discovery Rate\\GO & & Gene Ontology\\GSEA & & Gene Set Enrichment Analysis.\\R$_s$ & & Spearman correlation coefficient\\\end{tabular}%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%\newpage\bibliographystyle{plainnat}\begin{thebibliography}{4}
\providecommand{\natexlab}[1]{#1}
\providecommand{\url}[1]{\texttt{#1}}
\expandafter\ifx\csname urlstyle\endcsname\relax
  \providecommand{\doi}[1]{doi: #1}\else
  \providecommand{\doi}{doi: \begingroup \urlstyle{rm}\Url}\fi

\bibitem[Barretina et~al.(2012)Barretina, Caponigro, Stransky, Venkatesan,
  Margolin, Kim, Wilson, Leh{\'a}r, Kryukov, Sonkin, Reddy, Liu, Murray,
  Berger, Monahan, Morais, Meltzer, Korejwa, Jan{\'e}-Valbuena, Mapa, Thibault,
  Bric-Furlong, Raman, Shipway, Engels, Cheng, Yu, Yu, Aspesi, de~Silva,
  Jagtap, Jones, Wang, Hatton, Palescandolo, Gupta, Mahan, Sougnez, Onofrio,
  Liefeld, MacConaill, Winckler, Reich, Li, Mesirov, Gabriel, Getz, Ardlie,
  Chan, Myer, Weber, Porter, Warmuth, Finan, Harris, Meyerson, Golub,
  Morrissey, Sellers, Schlegel, and Garraway]{Barretina:2012fp}
Jordi Barretina, Giordano Caponigro, Nicolas Stransky, Kavitha Venkatesan,
  Adam~A Margolin, Sungjoon Kim, Christopher~J Wilson, Joseph Leh{\'a}r,
  Gregory~V Kryukov, Dmitriy Sonkin, Anupama Reddy, Manway Liu, Lauren Murray,
  Michael~F Berger, John~E Monahan, Paula Morais, Jodi Meltzer, Adam Korejwa,
  Judit Jan{\'e}-Valbuena, Felipa~A Mapa, Joseph Thibault, Eva Bric-Furlong,
  Pichai Raman, Aaron Shipway, Ingo~H Engels, Jill Cheng, Guoying~K Yu, Jianjun
  Yu, Peter Aspesi, Melanie de~Silva, Kalpana Jagtap, Michael~D Jones, Li~Wang,
  Charles Hatton, Emanuele Palescandolo, Supriya Gupta, Scott Mahan, Carrie
  Sougnez, Robert~C Onofrio, Ted Liefeld, Laura MacConaill, Wendy Winckler,
  Michael Reich, Nanxin Li, Jill~P. Mesirov, Stacey~B Gabriel, Gad Getz,
  Kristin Ardlie, Vivien Chan, Vic~E Myer, Barbara~L Weber, Jeff Porter, Markus
  Warmuth, Peter Finan, Jennifer~L Harris, Matthew Meyerson, Todd~R. Golub,
  Michael~P Morrissey, William~R Sellers, Robert Schlegel, and Levi~A.
  Garraway.
\newblock {The Cancer Cell Line Encyclopedia enables predictive modelling of
  anticancer drug sensitivity}.
\newblock \emph{Nature}, 483\penalty0 (7391):\penalty0 603--607, March 2012.

\bibitem[Garnett et~al.(2012)Garnett, Edelman, Heidorn, Greenman, Dastur, Lau,
  Greninger, Thompson, Luo, Soares, Liu, Iorio, Surdez, Chen, Milano, Bignell,
  Tam, Davies, Stevenson, Barthorpe, Lutz, Kogera, Lawrence, McLaren-Douglas,
  Mitropoulos, Mironenko, Thi, Richardson, Zhou, Jewitt, Zhang, O'Brien,
  Boisvert, Price, Hur, Yang, Deng, Butler, Choi, Chang, Baselga, Stamenkovic,
  Engelman, Sharma, Delattre, Saez-Rodriguez, Gray, Settleman, Futreal, Haber,
  Stratton, Ramaswamy, McDermott, and Benes]{Garnett:2012fc}
Mathew~J Garnett, Elena~J Edelman, Sonja~J Heidorn, Chris~D Greenman, Anahita
  Dastur, King~Wai Lau, Patricia Greninger, I~Richard Thompson, Xi~Luo, Jorge
  Soares, Qingsong Liu, Francesco Iorio, Didier Surdez, Li~Chen, Randy~J
  Milano, Graham~R Bignell, Ah~T Tam, Helen Davies, Jesse~A Stevenson, Syd
  Barthorpe, Stephen~R Lutz, Fiona Kogera, Karl Lawrence, Anne McLaren-Douglas,
  Xeni Mitropoulos, Tatiana Mironenko, Helen Thi, Laura Richardson, Wenjun
  Zhou, Frances Jewitt, Tinghu Zhang, Patrick O'Brien, Jessica~L Boisvert,
  Stacey Price, Wooyoung Hur, Wanjuan Yang, Xianming Deng, Adam Butler,
  Hwan~Geun Choi, Jae~Won Chang, Jose Baselga, Ivan Stamenkovic, Jeffrey~A
  Engelman, Sreenath~V Sharma, Olivier Delattre, Julio Saez-Rodriguez,
  Nathanael~S Gray, Jeffrey Settleman, P~Andrew Futreal, Daniel~A Haber,
  Michael~R Stratton, Sridhar Ramaswamy, Ultan McDermott, and Cyril~H Benes.
\newblock {Systematic identification of genomic markers of drug sensitivity in
  cancer cells.}
\newblock \emph{Nature}, 483\penalty0 (7391):\penalty0 570--575, March 2012.

\bibitem[Greshock et~al.(2010)Greshock, Bachman, Degenhardt, Jing, Wen,
  Eastman, McNeil, Moy, Wegrzyn, Auger, Hardwicke, and
  Wooster]{Greshock:2010hf}
J~Greshock, K~E Bachman, Y~Y Degenhardt, J~Jing, Y~H Wen, S~Eastman, E~McNeil,
  C~Moy, R~Wegrzyn, K~Auger, M~A Hardwicke, and R~Wooster.
\newblock {Molecular Target Class Is Predictive of In vitro Response Profile}.
\newblock \emph{Cancer Research}, 70\penalty0 (9):\penalty0 3677--3686, April
  2010.

\bibitem[Subramanian et~al.(2005)Subramanian, Tamayo, Mootha, Mukherjee, Ebert,
  Gillette, Paulovich, Pomeroy, Golub, Lander, and
  Mesirov]{subramanian2005gene}
Aravind Subramanian, Pablo Tamayo, Vamsi~K. Mootha, Sayan Mukherjee,
  Benjamin~L. Ebert, Michael~A. Gillette, Amanda Paulovich, Scott~L. Pomeroy,
  Todd~R. Golub, Eric~S. Lander, and Jill~P. Mesirov.
\newblock {Gene set enrichment analysis: A knowledge-based approach for
  interpreting genome-wide expression profiles}.
\newblock \emph{Proceedings of the National Academy of Sciences of the United
  States of America}, 102\penalty0 (43):\penalty0 15545--15550, 2005.

\end{thebibliography}
\end{document}